runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/grpcclient/GrpcGetWorkerMetadataStream.java

```diff
@@ -0,0 +1,285 @@
+            1, // logEveryNStreamFailures
+            jobHeader,
+            metadataVersion,
+            new ThrottleTimer(),
+            serverMappingUpdater);
+    testStream.fullyInitializeRequestObserver();
+    testStream.startStream();
+    return testStream;
+  }
+
+  /**
+   * Because {@link AbstractWindmillStream.ResponseObserver} is a non-static inner class, it (and
+   * its members) cannot be referenced until the constructor finishes executing and the object is
+   * completely instantiated. We gate this from being null by making the constructor private and
+   * exposing a static method so that an instance of {@link GrpcGetWorkerMetadataStream} will be
+   * created in the correct state.
+   */
+  private void fullyInitializeRequestObserver() {
+    getWorkMetadataStreamRequestObserver.init(
+        AbstractWindmillStream.ResponseObserver::new,
+        () -> {
+          synchronized (GrpcGetWorkerMetadataStream.this) {
+            streamClosed.set(true);
+          }
+        });
+  }
+
+  @Override
+  protected StreamObserver<WorkerMetadataRequest> createNewRequestObserver() {
+    // we do not need to re-initiate the requestObserver here, since every
+    // GetWorkMetadataStreamRequestObserver.onNext() starts a new server stream.
+    return getWorkMetadataStreamRequestObserver;
+  }
+
+  @Override
+  protected void onResponse(WorkerMetadataResponse response) {
+    synchronized (this) {
+      // If the metadataVersion newer or the same as the one in the response, the response data is
+      // stale, and we do not want to do anything.
+      long currentMetadataVersion = metadataVersion.get();
+      if (currentMetadataVersion >= response.getMetadataVersion()) {
+        LOG.info(
+            "Received metadata version={}; Current metadata version={}. "
+                + "Skipping update because received stale metadata",
+            response.getMetadataVersion(),
+            currentMetadataVersion);
+        return;
+      }
+      metadataVersion.set(response.getMetadataVersion());
+    }
+    ImmutableMap<String, String> globalDataServers =
+        response.getGlobalDataEndpointsMap().entrySet().stream()
+            .collect(
+                toImmutableMap(
+                    Map.Entry::getKey, // global data key
+                    endpoint -> endpoint.getValue().getEndpoint()));
+
+    ImmutableList<String> windmillServers =
+        response.getDirectPathEndpointsList().stream()
+            .map(WorkerMetadataResponse.Endpoint::getEndpoint)
+            .collect(toImmutableList());
+
+    serverMappingUpdater.accept(
+        WindmillEndpoints.builder()
+            .setGlobalDataServers(globalDataServers)
+            .setWindmillServers(windmillServers)
+            .build());
+  }
+
+  @Override
+  protected synchronized void onNewStream() {
+    send(workerMetadataRequest);
+  }
+
+  @Override
+  protected void startThrottleTimer() {
+    getWorkerMetadataThrottleTimer.start();
```

Comment:
@scwhittle are these throttle timers for client side throttling? if yes, i don't think we need them here.

Reply:
changed to bidi stream, so we will need this